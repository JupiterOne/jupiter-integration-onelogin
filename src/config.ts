import {
  IntegrationExecutionContext,
  IntegrationValidationError,
  IntegrationInstanceConfigFieldMap,
  IntegrationInstanceConfig,
} from '@jupiterone/integration-sdk-core';
import { createAPIClient } from './client';

/**
 * A type describing the configuration fields required to execute the
 * integration for a specific account in the data provider.
 *
 * When executing the integration in a development environment, these values may
 * be provided in a `.env` file with environment variables. For example:
 *
 * - `CLIENT_ID=123` becomes `instance.config.clientId = '123'`
 * - `CLIENT_SECRET=abc` becomes `instance.config.clientSecret = 'abc'`
 *
 * Environment variables are NOT used when the integration is executing in a
 * managed environment. For example, in JupiterOne, users configure
 * `instance.config` in a UI.
 */
export const instanceConfigFields: IntegrationInstanceConfigFieldMap = {
  clientId: {
    type: 'string',
  },
  clientSecret: {
    type: 'string',
    mask: true,
  },
  orgUrl: {
    type: 'string',
  },
};

/**
 * Properties provided by the `IntegrationInstance.config`. This reflects the
 * same properties defined by `instanceConfigFields`.
 */
export interface IntegrationConfig extends IntegrationInstanceConfig {
  /**
   * The Onelogin client ID used to authenticate requests.
   */
  clientId: string;

  /**
   * The Onelogin client secret used to authenticate requests.
   */
  clientSecret: string;

  /**
   * The Onelogin organization URL. Only used to create a weblink to the account.
   */
  orgUrl: string;

  /**
   * The Onelogin account name. Optional. Used to create name of account entity in J1 graph.
   */
  accountName?: string;
}

export async function validateInvocation(
  context: IntegrationExecutionContext<IntegrationConfig>,
) {
  const { config } = context.instance;

  if (!config.clientId || !config.clientSecret) {
    throw new IntegrationValidationError(
      'Config requires all of {clientId, clientSecret}',
    );
  }

  //checks for orgUrl
  if (/https?:\/\//.test(config.orgUrl)) {
    throw new IntegrationValidationError(
      'Config {orgUrl} should not have https:// prepended',
    );
  }

  const splitter = config.orgUrl.split('.');
  if (!(splitter[1] === 'onelogin')) {
    throw new IntegrationValidationError(
      'Problem with config {orgUrl}. Should be {YOURDOMAIN}.onelogin.com',
    );
  }

  const apiClient = createAPIClient(config, context.logger);
  await apiClient.verifyAuthentication();
}
