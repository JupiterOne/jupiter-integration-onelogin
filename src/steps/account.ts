import {
  IntegrationStep,
  IntegrationStepExecutionContext,
  createDirectRelationship,
  RelationshipClass,
} from '@jupiterone/integration-sdk-core';

import { IntegrationConfig } from '../config';
import { createAccountEntity, getServiceEntities } from '../converters';
import {
  ACCOUNT_ENTITY_TYPE,
  ACCOUNT_ENTITY_CLASS,
  SERVICE_ENTITY_CLASS,
  SERVICE_ENTITY_TYPE,
  ACCOUNT_SERVICE_RELATIONSHIP_TYPE,
} from '../jupiterone';

export const DATA_ACCOUNT_ENTITY = 'DATA_ACCOUNT_ENTITY';

export async function fetchAccountDetails({
  instance,
  jobState,
}: IntegrationStepExecutionContext<IntegrationConfig>) {
  const account = {
    id: instance.id,
    name: instance.config.accountName || instance.name,
    orgUrl: instance.config.orgUrl,
  };
  const accountEntity = await jobState.addEntity(createAccountEntity(account));
  await jobState.setData(DATA_ACCOUNT_ENTITY, accountEntity);

  const serviceEntities = await jobState.addEntities(
    getServiceEntities(account.name),
  );
  for (const serviceEntity of serviceEntities) {
    await jobState.addRelationship(
      createDirectRelationship({
        _class: RelationshipClass.HAS,
        from: accountEntity,
        to: serviceEntity,
      }),
    );
  }
}

export const accountSteps: IntegrationStep<IntegrationConfig>[] = [
  {
    id: 'fetch-account',
    name: 'Fetch Account Details',
    entities: [
      {
        resourceName: 'Onelogin Account',
        _type: ACCOUNT_ENTITY_TYPE,
        _class: ACCOUNT_ENTITY_CLASS,
      },
      {
        resourceName: 'Onelogin Service',
        _type: SERVICE_ENTITY_TYPE,
        _class: SERVICE_ENTITY_CLASS,
      },
    ],
    relationships: [
      {
        _type: ACCOUNT_SERVICE_RELATIONSHIP_TYPE,
        _class: RelationshipClass.HAS,
        sourceType: ACCOUNT_ENTITY_TYPE,
        targetType: SERVICE_ENTITY_TYPE,
      },
    ],
    dependsOn: [],
    executionHandler: fetchAccountDetails,
  },
];
