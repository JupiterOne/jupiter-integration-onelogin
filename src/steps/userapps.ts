import {
  createDirectRelationship,
  IntegrationStep,
  IntegrationStepExecutionContext,
  RelationshipClass,
  IntegrationMissingKeyError,
} from '@jupiterone/integration-sdk-core';

import { createAPIClient } from '../client';
import { IntegrationConfig } from '../config';
import { createPersonalAppEntity } from '../converters';
import {
  UserEntity,
  AppEntity,
  IdEntityMap,
  PERSONAL_APP_ENTITY_TYPE,
  PERSONAL_APP_ENTITY_CLASS,
  USER_PERSONAL_APP_RELATIONSHIP_TYPE,
  USER_ENTITY_TYPE,
  USER_APP_RELATIONSHIP_TYPE,
  APP_ENTITY_TYPE,
} from '../jupiterone';

export async function fetchUserApps({
  instance,
  jobState,
  logger,
}: IntegrationStepExecutionContext<IntegrationConfig>) {
  const apiClient = createAPIClient(instance.config, logger);

  const userEntities = await jobState.getData<UserEntity[]>('USER_ARRAY');

  if (!userEntities) {
    throw new IntegrationMissingKeyError(
      `Expected to find User entity array in jobState.`,
    );
  }

  const appByIdMap = await jobState.getData<IdEntityMap<AppEntity>>(
    'APPLICATION_BY_ID_MAP',
  );

  if (!appByIdMap) {
    throw new IntegrationMissingKeyError(
      `Expected to find appByIdMap in jobState.`,
    );
  }

  for (const userEntity of userEntities) {
    await apiClient.iterateUserApps(userEntity.id, async (userApp) => {
      if (appByIdMap[userApp.id]) {
        //in this case, it is an organization app
        await jobState.addRelationship(
          createDirectRelationship({
            _class: RelationshipClass.ASSIGNED,
            from: userEntity,
            to: appByIdMap[userApp.id],
          }),
        );
      }

      //documentation and experiments suggest that a user's personal apps
      //are not reported by the API, but just in case they are:
      if (!appByIdMap[userApp.id] && userApp.personal === true) {
        const personalAppEntity = await jobState.addEntity(
          createPersonalAppEntity(userApp),
        );
        await jobState.addRelationship(
          createDirectRelationship({
            _class: RelationshipClass.HAS,
            from: userEntity,
            to: personalAppEntity,
          }),
        );
      }
    });
  }
}

export const userAppSteps: IntegrationStep<IntegrationConfig>[] = [
  {
    id: 'fetch-userapps',
    name: 'Fetch User Apps',
    entities: [
      {
        resourceName: 'Onelogin Personal Application',
        _type: PERSONAL_APP_ENTITY_TYPE,
        _class: PERSONAL_APP_ENTITY_CLASS,
      },
    ],
    relationships: [
      {
        _type: USER_APP_RELATIONSHIP_TYPE,
        _class: RelationshipClass.ASSIGNED,
        sourceType: USER_ENTITY_TYPE,
        targetType: APP_ENTITY_TYPE,
      },
      {
        _type: USER_PERSONAL_APP_RELATIONSHIP_TYPE,
        _class: RelationshipClass.HAS,
        sourceType: USER_ENTITY_TYPE,
        targetType: PERSONAL_APP_ENTITY_TYPE,
      },
    ],
    dependsOn: ['fetch-users', 'fetch-applications'],
    executionHandler: fetchUserApps,
  },
];
