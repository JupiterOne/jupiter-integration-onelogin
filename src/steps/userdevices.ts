import {
  createDirectRelationship,
  IntegrationStep,
  IntegrationStepExecutionContext,
  RelationshipClass,
  IntegrationMissingKeyError,
} from '@jupiterone/integration-sdk-core';

import { createAPIClient } from '../client';
import { IntegrationConfig } from '../config';
import { createPersonalDeviceEntity } from '../converters';
import {
  UserEntity,
  PERSONAL_DEVICE_ENTITY_TYPE,
  PERSONAL_DEVICE_ENTITY_CLASS,
  USER_ENTITY_TYPE,
  USER_PERSONAL_DEVICE_RELATIONSHIP_TYPE,
} from '../jupiterone';

export async function fetchUserDevices({
  instance,
  jobState,
  logger,
}: IntegrationStepExecutionContext<IntegrationConfig>) {
  const apiClient = createAPIClient(instance.config, logger);

  const userEntities = await jobState.getData<UserEntity[]>('USER_ARRAY');

  if (!userEntities) {
    throw new IntegrationMissingKeyError(
      `Expected to find User entity array in jobState.`,
    );
  }

  for (const userEntity of userEntities) {
    await apiClient.iterateUserDevices(userEntity.id, async (device) => {
      const deviceEntity = await jobState.addEntity(
        createPersonalDeviceEntity(device),
      );

      await jobState.addRelationship(
        createDirectRelationship({
          _class: RelationshipClass.ASSIGNED,
          from: userEntity,
          to: deviceEntity,
        }),
      );
    });
  }
}

export const userDeviceSteps: IntegrationStep<IntegrationConfig>[] = [
  {
    id: 'fetch-userdevices',
    name: 'Fetch User Devices',
    entities: [
      {
        resourceName: 'Onelogin Personal Device',
        _type: PERSONAL_DEVICE_ENTITY_TYPE,
        _class: PERSONAL_DEVICE_ENTITY_CLASS,
      },
    ],
    relationships: [
      {
        _type: USER_PERSONAL_DEVICE_RELATIONSHIP_TYPE,
        _class: RelationshipClass.ASSIGNED,
        sourceType: USER_ENTITY_TYPE,
        targetType: PERSONAL_DEVICE_ENTITY_TYPE,
      },
    ],
    dependsOn: ['fetch-users'],
    executionHandler: fetchUserDevices,
  },
];
