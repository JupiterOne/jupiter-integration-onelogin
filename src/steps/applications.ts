import {
  createDirectRelationship,
  Entity,
  IntegrationStep,
  IntegrationStepExecutionContext,
  RelationshipClass,
  IntegrationMissingKeyError,
} from '@jupiterone/integration-sdk-core';

import { createAPIClient } from '../client';
import { IntegrationConfig } from '../config';
import { createAppEntity } from '../converters';
import { DATA_ACCOUNT_ENTITY } from './account';
import {
  ACCOUNT_ENTITY_TYPE,
  ACCOUNT_APP_RELATIONSHIP_TYPE,
  RoleEntity,
  APP_ENTITY_CLASS,
  APP_ENTITY_TYPE,
} from '../jupiterone';

export async function fetchApplications({
  instance,
  jobState,
  logger,
}: IntegrationStepExecutionContext<IntegrationConfig>) {
  const apiClient = createAPIClient(instance.config, logger);

  const accountEntity = (await jobState.getData(DATA_ACCOUNT_ENTITY)) as Entity;

  if (!accountEntity) {
    throw new IntegrationMissingKeyError(
      `Expected to find Account entity in jobState.`,
    );
  }

  //for use later in other steps
  const appByIdMap = {};

  await apiClient.iterateApplications(async (app) => {
    const applicationEntity = (await jobState.addEntity(
      createAppEntity(app),
    )) as RoleEntity;

    await jobState.addRelationship(
      createDirectRelationship({
        _class: RelationshipClass.HAS,
        from: accountEntity,
        to: applicationEntity,
      }),
    );

    appByIdMap[app.id] = applicationEntity;
  });

  await jobState.setData('APPLICATION_BY_ID_MAP', appByIdMap);
}

export const applicationSteps: IntegrationStep<IntegrationConfig>[] = [
  {
    id: 'fetch-applications',
    name: 'Fetch Applications',
    entities: [
      {
        resourceName: 'Onelogin Application',
        _type: APP_ENTITY_TYPE,
        _class: APP_ENTITY_CLASS,
      },
    ],
    relationships: [
      {
        _type: ACCOUNT_APP_RELATIONSHIP_TYPE,
        _class: RelationshipClass.HAS,
        sourceType: ACCOUNT_ENTITY_TYPE,
        targetType: APP_ENTITY_TYPE,
      },
    ],
    dependsOn: ['fetch-account'],
    executionHandler: fetchApplications,
  },
];
