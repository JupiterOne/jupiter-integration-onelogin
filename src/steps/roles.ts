import {
  createDirectRelationship,
  Entity,
  IntegrationStep,
  IntegrationStepExecutionContext,
  RelationshipClass,
  IntegrationMissingKeyError,
} from '@jupiterone/integration-sdk-core';

import { createAPIClient } from '../client';
import { IntegrationConfig } from '../config';
import { createRoleEntity } from '../converters';
import { DATA_ACCOUNT_ENTITY } from './account';
import {
  ACCOUNT_ENTITY_TYPE,
  ACCOUNT_ROLE_RELATIONSHIP_TYPE,
  RoleEntity,
  ROLE_ENTITY_CLASS,
  ROLE_ENTITY_TYPE,
} from '../jupiterone';

export async function fetchRoles({
  instance,
  jobState,
  logger,
}: IntegrationStepExecutionContext<IntegrationConfig>) {
  const apiClient = createAPIClient(instance.config, logger);

  const accountEntity = (await jobState.getData(DATA_ACCOUNT_ENTITY)) as Entity;

  if (!accountEntity) {
    throw new IntegrationMissingKeyError(
      `Expected to find Account entity in jobState.`,
    );
  }

  //for use later in other steps
  const roleByIdMap = {};

  await apiClient.iterateRoles(async (role) => {
    const roleEntity = (await jobState.addEntity(
      createRoleEntity(role),
    )) as RoleEntity;

    await jobState.addRelationship(
      createDirectRelationship({
        _class: RelationshipClass.HAS,
        from: accountEntity,
        to: roleEntity,
      }),
    );

    roleByIdMap[role.id] = roleEntity;
  });

  await jobState.setData('ROLE_BY_ID_MAP', roleByIdMap);
}

export const roleSteps: IntegrationStep<IntegrationConfig>[] = [
  {
    id: 'fetch-roles',
    name: 'Fetch Roles',
    entities: [
      {
        resourceName: 'Onelogin Role',
        _type: ROLE_ENTITY_TYPE,
        _class: ROLE_ENTITY_CLASS,
      },
    ],
    relationships: [
      {
        _type: ACCOUNT_ROLE_RELATIONSHIP_TYPE,
        _class: RelationshipClass.HAS,
        sourceType: ACCOUNT_ENTITY_TYPE,
        targetType: ROLE_ENTITY_TYPE,
      },
    ],
    dependsOn: ['fetch-account'],
    executionHandler: fetchRoles,
  },
];
