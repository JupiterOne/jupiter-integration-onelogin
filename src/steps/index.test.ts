import {
  createMockStepExecutionContext,
  Recording,
} from '@jupiterone/integration-sdk-testing';

import { IntegrationConfig } from '../config';
import { fetchUsers } from './users';
import { fetchGroups } from './groups';
import { fetchRoles } from './roles';
import { fetchApplications } from './applications';
import { fetchUserApps } from './userapps';
import { fetchUserDevices } from './userdevices';
import { fetchAccountDetails } from './account';

import { integrationConfig } from '../../test/config';
import { setupOneloginRecording } from '../../test/recording';

let recording: Recording;

afterEach(async () => {
  await recording.stop();
});

test('should collect data', async () => {
  recording = setupOneloginRecording({
    directory: __dirname,
    name: 'steps',
  });

  const context = createMockStepExecutionContext<IntegrationConfig>({
    instanceConfig: integrationConfig,
  });

  // Simulates dependency graph execution.
  // See https://github.com/JupiterOne/sdk/issues/262.
  await fetchAccountDetails(context);
  await fetchGroups(context);
  await fetchRoles(context);
  await fetchUsers(context);
  await fetchApplications(context);
  await fetchUserApps(context);
  await fetchUserDevices(context);

  // Review snapshot, failure is a regression
  expect({
    numCollectedEntities: context.jobState.collectedEntities.length,
    numCollectedRelationships: context.jobState.collectedRelationships.length,
    collectedEntities: context.jobState.collectedEntities,
    collectedRelationships: context.jobState.collectedRelationships,
    encounteredTypes: context.jobState.encounteredTypes,
  }).toMatchSnapshot();

  const accounts = context.jobState.collectedEntities.filter((e) =>
    e._class.includes('Account'),
  );
  expect(accounts.length).toBeGreaterThan(0);
  expect(accounts).toMatchGraphObjectSchema({
    _class: ['Account'],
    schema: {
      additionalProperties: true,
      properties: {
        _type: { const: 'onelogin_account' },
        name: { type: 'string' },
        displayName: { type: 'string' },
        webLink: { type: 'string', format: 'url' },
        _rawData: {
          type: 'array',
          items: { type: 'object' },
        },
      },
      required: ['name', 'displayName', 'webLink'],
    },
  });

  const groups = context.jobState.collectedEntities.filter((e) =>
    e._class.includes('UserGroup'),
  );
  expect(groups.length).toBeGreaterThan(0);
  expect(groups).toMatchGraphObjectSchema({
    _class: ['UserGroup'],
    schema: {
      additionalProperties: true,
      properties: {
        _type: { const: 'onelogin_group' },
        name: { type: 'string' },
        _rawData: {
          type: 'array',
          items: { type: 'object' },
        },
      },
      required: ['name', 'id'],
    },
  });

  const roles = context.jobState.collectedEntities.filter((e) =>
    e._class.includes('AccessRole'),
  );
  expect(roles.length).toBeGreaterThan(0);
  expect(roles).toMatchGraphObjectSchema({
    _class: ['AccessRole'],
    schema: {
      additionalProperties: true,
      properties: {
        _type: { const: 'onelogin_role' },
        name: { type: 'string' },
        _rawData: {
          type: 'array',
          items: { type: 'object' },
        },
      },
      required: ['name', 'id'],
    },
  });

  const users = context.jobState.collectedEntities.filter((e) =>
    e._class.includes('User'),
  );
  expect(users.length).toBeGreaterThan(0);
  expect(users).toMatchGraphObjectSchema({
    _class: ['User'],
    schema: {
      additionalProperties: true,
      properties: {
        _type: { const: 'onelogin_user' },
        name: { type: 'string' },
        displayName: { type: 'string' },
        email: { type: 'string' },
        _rawData: {
          type: 'array',
          items: { type: 'object' },
        },
      },
      required: ['name', 'displayName', 'email'],
    },
  });

  const applications = context.jobState.collectedEntities.filter((e) =>
    e._class.includes('Application'),
  );
  expect(applications.length).toBeGreaterThan(0);
  expect(applications).toMatchGraphObjectSchema({
    _class: ['Application'],
    schema: {
      additionalProperties: true,
      properties: {
        _type: { const: 'onelogin_application' },
        name: { type: 'string' },
        _rawData: {
          type: 'array',
          items: { type: 'object' },
        },
      },
      required: ['name', 'id'],
    },
  });
});
