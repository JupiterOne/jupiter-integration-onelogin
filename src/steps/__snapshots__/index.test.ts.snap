// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should collect data 1`] = `
Object {
  "collectedEntities": Array [
    Object {
      "_class": Array [
        "Account",
      ],
      "_key": "onelogin_account_local-integration-instance",
      "_type": "onelogin_account",
      "displayName": "Local Integration",
      "name": "Local Integration",
      "webLink": "https://orgUrl.com",
    },
    Object {
      "_class": Array [
        "Service",
        "Control",
      ],
      "_key": "onelogin:sso:Local Integration",
      "_type": "onelogin_service",
      "category": "security",
      "controlDomain": "identity-access",
      "displayName": "Onelogin SSO",
      "function": "sso",
      "name": "SSO",
    },
    Object {
      "_class": Array [
        "Service",
        "Control",
      ],
      "_key": "onelogin:mfa:Local Integration",
      "_type": "onelogin_service",
      "category": "security",
      "controlDomain": "identity-access",
      "displayName": "Onelogin MFA",
      "function": "mfa",
      "name": "MFA",
    },
    Object {
      "_class": Array [
        "UserGroup",
      ],
      "_key": "onelogin_group_477661",
      "_type": "onelogin_group",
      "displayName": "Getting the Band Together",
      "id": "477661",
      "name": "Getting the Band Together",
      "reference": null,
    },
    Object {
      "_class": Array [
        "AccessRole",
      ],
      "_key": "onelogin_role_428308",
      "_type": "onelogin_role",
      "displayName": "Default",
      "id": "428308",
      "name": "Default",
    },
    Object {
      "_class": Array [
        "AccessRole",
      ],
      "_key": "onelogin_role_439868",
      "_type": "onelogin_role",
      "displayName": "Rock Star",
      "id": "439868",
      "name": "Rock Star",
    },
    Object {
      "_class": Array [
        "AccessRole",
      ],
      "_key": "onelogin_role_439869",
      "_type": "onelogin_role",
      "displayName": "Ninja",
      "id": "439869",
      "name": "Ninja",
    },
    Object {
      "_class": Array [
        "AccessRole",
      ],
      "_key": "onelogin_role_439870",
      "_type": "onelogin_role",
      "displayName": "Cryptozoologist",
      "id": "439870",
      "name": "Cryptozoologist",
    },
    Object {
      "_class": Array [
        "User",
      ],
      "_key": "onelogin_user_132422371",
      "_type": "onelogin_user",
      "activatedAt": 1619644257207,
      "comment": "",
      "company": "",
      "createdAt": 1619644257210,
      "customAttributes.customUserField": "custom-user-value",
      "department": "",
      "directoryId": null,
      "displayName": "Examplea Person",
      "distinguishedName": null,
      "email": "exampleperson1@jupiterone.com",
      "externalId": null,
      "firstname": "Examplea",
      "groupId": 0,
      "id": "132422371",
      "invalidLoginAttempts": 0,
      "invitationSentAt": 1619644257187,
      "lastLogin": 1623866368297,
      "lastname": "Person",
      "localeCode": null,
      "lockedUntil": null,
      "managerAdId": null,
      "managerUserId": null,
      "memberOf": null,
      "name": "Examplea Person",
      "openidName": "examplea.person",
      "passwordChangedAt": 1623866359915,
      "phone": "",
      "preferredLocaleCode": null,
      "samaccountname": null,
      "state": 1,
      "status": "1",
      "title": "",
      "trustedIdpId": null,
      "updatedAt": 1623866368842,
      "username": "",
      "userprincipalname": null,
    },
    Object {
      "_class": Array [
        "User",
      ],
      "_key": "onelogin_user_133771160",
      "_type": "onelogin_user",
      "activatedAt": 1620741537421,
      "comment": "",
      "company": "",
      "createdAt": 1620741537430,
      "customAttributes.customUserField": "",
      "department": "",
      "directoryId": null,
      "displayName": "Exampleb Person",
      "distinguishedName": null,
      "email": "exampleperson2@jupiterone.com",
      "externalId": null,
      "firstname": "Exampleb",
      "groupId": 477661,
      "id": "133771160",
      "invalidLoginAttempts": 0,
      "invitationSentAt": undefined,
      "lastLogin": undefined,
      "lastname": "Person",
      "localeCode": null,
      "lockedUntil": null,
      "managerAdId": null,
      "managerUserId": null,
      "memberOf": null,
      "name": "Exampleb Person",
      "openidName": "exampleb.person",
      "passwordChangedAt": undefined,
      "phone": "",
      "preferredLocaleCode": null,
      "samaccountname": null,
      "state": 1,
      "status": "1",
      "title": "",
      "trustedIdpId": null,
      "updatedAt": 1623867217199,
      "username": "",
      "userprincipalname": null,
    },
    Object {
      "_class": Array [
        "User",
      ],
      "_key": "onelogin_user_137729105",
      "_type": "onelogin_user",
      "activatedAt": 1623866437300,
      "comment": "",
      "company": "",
      "createdAt": 1623866437485,
      "customAttributes.customUserField": "",
      "department": "",
      "directoryId": null,
      "displayName": "Examplec Person",
      "distinguishedName": null,
      "email": "exampleperson3@gmail.com",
      "externalId": null,
      "firstname": "Examplec",
      "groupId": 477661,
      "id": "137729105",
      "invalidLoginAttempts": 0,
      "invitationSentAt": 1623866452170,
      "lastLogin": 1623960908941,
      "lastname": "Person",
      "localeCode": null,
      "lockedUntil": null,
      "managerAdId": null,
      "managerUserId": null,
      "memberOf": null,
      "name": "Examplec Person",
      "openidName": "examplec.person",
      "passwordChangedAt": 1623866732455,
      "phone": "",
      "preferredLocaleCode": null,
      "samaccountname": null,
      "state": 1,
      "status": "1",
      "title": "",
      "trustedIdpId": null,
      "updatedAt": 1623960909465,
      "username": "",
      "userprincipalname": null,
    },
    Object {
      "_class": Array [
        "Application",
      ],
      "_icon": "https://s3.amazonaws.com/onelogin-assets/images/icons/doordash.png",
      "_key": "onelogin_application_1475668",
      "_type": "onelogin_application",
      "connectorId": 68269,
      "displayName": "DoorDash",
      "extension": true,
      "id": "1475668",
      "name": "DoorDash",
      "provisioning": false,
      "visible": true,
    },
  ],
  "collectedRelationships": Array [
    Object {
      "_class": "HAS",
      "_fromEntityKey": "onelogin_account_local-integration-instance",
      "_key": "onelogin_account_local-integration-instance|has|onelogin:sso:Local Integration",
      "_toEntityKey": "onelogin:sso:Local Integration",
      "_type": "onelogin_account_has_service",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "onelogin_account_local-integration-instance",
      "_key": "onelogin_account_local-integration-instance|has|onelogin:mfa:Local Integration",
      "_toEntityKey": "onelogin:mfa:Local Integration",
      "_type": "onelogin_account_has_service",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "onelogin_account_local-integration-instance",
      "_key": "onelogin_account_local-integration-instance|has|onelogin_group_477661",
      "_toEntityKey": "onelogin_group_477661",
      "_type": "onelogin_account_has_group",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "onelogin_account_local-integration-instance",
      "_key": "onelogin_account_local-integration-instance|has|onelogin_role_428308",
      "_toEntityKey": "onelogin_role_428308",
      "_type": "onelogin_account_has_role",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "onelogin_account_local-integration-instance",
      "_key": "onelogin_account_local-integration-instance|has|onelogin_role_439868",
      "_toEntityKey": "onelogin_role_439868",
      "_type": "onelogin_account_has_role",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "onelogin_account_local-integration-instance",
      "_key": "onelogin_account_local-integration-instance|has|onelogin_role_439869",
      "_toEntityKey": "onelogin_role_439869",
      "_type": "onelogin_account_has_role",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "onelogin_account_local-integration-instance",
      "_key": "onelogin_account_local-integration-instance|has|onelogin_role_439870",
      "_toEntityKey": "onelogin_role_439870",
      "_type": "onelogin_account_has_role",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "onelogin_account_local-integration-instance",
      "_key": "onelogin_account_local-integration-instance|has|onelogin_user_132422371",
      "_toEntityKey": "onelogin_user_132422371",
      "_type": "onelogin_account_has_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "ASSIGNED",
      "_fromEntityKey": "onelogin_user_132422371",
      "_key": "onelogin_user_132422371|assigned|onelogin_role_439868",
      "_toEntityKey": "onelogin_role_439868",
      "_type": "onelogin_user_assigned_role",
      "displayName": "ASSIGNED",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "onelogin_account_local-integration-instance",
      "_key": "onelogin_account_local-integration-instance|has|onelogin_user_133771160",
      "_toEntityKey": "onelogin_user_133771160",
      "_type": "onelogin_account_has_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "onelogin_group_477661",
      "_key": "onelogin_group_477661|has|onelogin_user_133771160",
      "_toEntityKey": "onelogin_user_133771160",
      "_type": "onelogin_group_has_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "ASSIGNED",
      "_fromEntityKey": "onelogin_user_133771160",
      "_key": "onelogin_user_133771160|assigned|onelogin_group_477661",
      "_toEntityKey": "onelogin_group_477661",
      "_type": "onelogin_user_assigned_group",
      "displayName": "ASSIGNED",
    },
    Object {
      "_class": "ASSIGNED",
      "_fromEntityKey": "onelogin_user_133771160",
      "_key": "onelogin_user_133771160|assigned|onelogin_role_439868",
      "_toEntityKey": "onelogin_role_439868",
      "_type": "onelogin_user_assigned_role",
      "displayName": "ASSIGNED",
    },
    Object {
      "_class": "ASSIGNED",
      "_fromEntityKey": "onelogin_user_133771160",
      "_key": "onelogin_user_133771160|assigned|onelogin_role_439870",
      "_toEntityKey": "onelogin_role_439870",
      "_type": "onelogin_user_assigned_role",
      "displayName": "ASSIGNED",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "onelogin_account_local-integration-instance",
      "_key": "onelogin_account_local-integration-instance|has|onelogin_user_137729105",
      "_toEntityKey": "onelogin_user_137729105",
      "_type": "onelogin_account_has_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "onelogin_group_477661",
      "_key": "onelogin_group_477661|has|onelogin_user_137729105",
      "_toEntityKey": "onelogin_user_137729105",
      "_type": "onelogin_group_has_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "ASSIGNED",
      "_fromEntityKey": "onelogin_user_137729105",
      "_key": "onelogin_user_137729105|assigned|onelogin_group_477661",
      "_toEntityKey": "onelogin_group_477661",
      "_type": "onelogin_user_assigned_group",
      "displayName": "ASSIGNED",
    },
    Object {
      "_class": "ASSIGNED",
      "_fromEntityKey": "onelogin_user_137729105",
      "_key": "onelogin_user_137729105|assigned|onelogin_role_439869",
      "_toEntityKey": "onelogin_role_439869",
      "_type": "onelogin_user_assigned_role",
      "displayName": "ASSIGNED",
    },
    Object {
      "_class": "ASSIGNED",
      "_fromEntityKey": "onelogin_user_137729105",
      "_key": "onelogin_user_137729105|assigned|onelogin_role_439870",
      "_toEntityKey": "onelogin_role_439870",
      "_type": "onelogin_user_assigned_role",
      "displayName": "ASSIGNED",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "onelogin_account_local-integration-instance",
      "_key": "onelogin_account_local-integration-instance|has|onelogin_application_1475668",
      "_toEntityKey": "onelogin_application_1475668",
      "_type": "onelogin_account_has_application",
      "displayName": "HAS",
    },
    Object {
      "_class": "ASSIGNED",
      "_fromEntityKey": "onelogin_user_133771160",
      "_key": "onelogin_user_133771160|assigned|onelogin_application_1475668",
      "_toEntityKey": "onelogin_application_1475668",
      "_type": "onelogin_user_assigned_application",
      "displayName": "ASSIGNED",
    },
    Object {
      "_class": "ASSIGNED",
      "_fromEntityKey": "onelogin_user_137729105",
      "_key": "onelogin_user_137729105|assigned|onelogin_application_1475668",
      "_toEntityKey": "onelogin_application_1475668",
      "_type": "onelogin_user_assigned_application",
      "displayName": "ASSIGNED",
    },
  ],
  "encounteredTypes": Array [
    "onelogin_account",
    "onelogin_service",
    "onelogin_account_has_service",
    "onelogin_group",
    "onelogin_account_has_group",
    "onelogin_role",
    "onelogin_account_has_role",
    "onelogin_user",
    "onelogin_account_has_user",
    "onelogin_user_assigned_role",
    "onelogin_group_has_user",
    "onelogin_user_assigned_group",
    "onelogin_application",
    "onelogin_account_has_application",
    "onelogin_user_assigned_application",
  ],
  "numCollectedEntities": 12,
  "numCollectedRelationships": 22,
}
`;
